cpp/src/arrow/array/util.cc

```diff
@@ -669,11 +671,29 @@ class RepeatedArrayFactory {
   enable_if_base_binary<T, Status> Visit(const T&) {
     const std::shared_ptr<Buffer>& value = scalar<T>().value;
     std::shared_ptr<Buffer> values_buffer, offsets_buffer;
-    RETURN_NOT_OK(CreateBufferOf(value->data(), value->size(), &values_buffer));
     auto size = static_cast<typename T::offset_type>(value->size());
+
+    typename T::offset_type length_casted;
+    if (length_ < 0) {
+      return Status::Invalid("length cannot be negative: " + std::to_string(length_));
+    } else if (length_ > std::numeric_limits<typename T::offset_type>::max()) {
+      return Status::Invalid(
+          "length exceeds the maximum value of offset_type: " + std::to_string(length_) +
+          " is greater than " +
+          std::to_string(std::numeric_limits<typename T::offset_type>::max()));
```

Comment:
I have no problem having this check here.
BTW You can just use `Status::Invalid("some message ", length_, " some message ", number)` without using `to_string` and `+`

Reply:
On second thought, what if the input scalar an empty string ""? Can the length be larger than int32::max?

Reply:
It could be larger than in32::max if the string were empty. We should only apply this check to the product of `(length of repeated string) * (number of repeats)`.